File: android/app/src/main/java/ai/neurophone/NativeLib.java
// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: 2026 Jonathan D.A. Jewell
package ai.neurophone;

/**
 * JNI surface declarations for {@code ai.neurophone.NativeLib}.
 *
 * <p>This class carries <em>zero</em> logic: every method is a {@code native}
 * declaration that resolves directly to a {@code #[unsafe(no_mangle)]}
 * {@code extern "system" fn Java_ai_neurophone_NativeLib_*} export in
 * {@code crates/neurophone-android/src/lib.rs}. That crate is the real JNI
 * bridge implementation (issue #110, already merged on {@code main}); this
 * file only has to match its 11-method signature exactly so
 * {@code System.loadLibrary} resolution succeeds. See
 * {@code docs/migrations/JNI-SURFACE-AUDIT.adoc} for the authoritative
 * signature table.
 *
 * <p>Part of the Android Kotlin&rarr;Rust/Gossamer migration (epic #83,
 * sub-issue #109). Exempt from the estate Java/Kotlin ban via the
 * {@code android/**}{@code /src/**}{@code /*.java} carve-out in
 * {@code hyperpolymath/standards} {@code governance-reusable.yml} (RFC
 * {@code docs/migrations/RFC-ANDROID-KOTLIN-TO-RUST.adoc} Q1, standards#341).
 */
public final class NativeLib {

    static {
        // Cargo cdylib output for the `neurophone-android` package:
        // libneurophone_android.so (hyphens -> underscores).
        System.loadLibrary("neurophone_android");
    }

    private NativeLib() {
        // Static-methods-only holder; never instantiated.
    }

    /** {@code init(configJson: String?): Boolean} */
    public static native boolean init(String configJson);

    /** {@code start(): Boolean} */
    public static native boolean start();

    /** {@code stop()} */
    public static native void stop();

    /** {@code isRunning(): Boolean} */
    public static native boolean isRunning();

    /**
     * {@code processSensor(sensorType: Int, values: FloatArray, timestamp: Long,
     * accuracy: Int): Boolean}
     */
    public static native boolean processSensor(int sensorType, float[] values, long timestamp, int accuracy);

    /** {@code query(message: String, preferLocal: Boolean): String} */
    public static native String query(String message, boolean preferLocal);

    /** {@code queryLocal(message: String): String} — forces the local model. */
    public static native String queryLocal(String message);

    /** {@code queryClaude(message: String): String} — forces the cloud model. */
    public static native String queryClaude(String message);

    /** {@code getNeuralContext(): String} */
    public static native String getNeuralContext();

    /** {@code getState(): String} — the {@code SystemState} as JSON. */
    public static native String getState();

    /** {@code reset()} */
    public static native void reset();
}

File: android/app/src/main/java/ai/neurophone/NativeLib.kt
// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: 2025 Jonathan D.A. Jewell
package ai.neurophone

/**
 * Native library interface for the NeuroSymbolic system.
 * Connects Kotlin/Android to the Rust core via JNI.
 */
object NativeLib {

    init {
        System.loadLibrary("neurophone_android")
    }

    /**
     * Initialize the native system with optional JSON config
     */
    external fun init(configJson: String? = null): Boolean

    /**
     * Start the neural processing loop
     */
    external fun start(): Boolean

    /**
     * Stop the neural processing loop
     */
    external fun stop()

    /**
     * Process sensor data
     * @param sensorType Android sensor type constant
     * @param values Sensor values array
     * @param timestamp Event timestamp in nanoseconds
     * @param accuracy Sensor accuracy level
     */
    external fun processSensor(
        sensorType: Int,
        values: FloatArray,
        timestamp: Long,
        accuracy: Int
    ): Boolean

    /**
     * Query local LLM (Llama 3.2)
     */
    external fun queryLocal(message: String): String

    /**
     * Query Claude (cloud)
     */
    external fun queryClaude(message: String): String

    /**
     * Smart query - auto-selects local or cloud
     */
    external fun query(message: String, preferLocal: Boolean = true): String

    /**
     * Get current neural context as formatted string
     */
    external fun getNeuralContext(): String

    /**
     * Get system state as JSON
     */
    external fun getState(): String

    /**
     * Reset all neural components
     */
    external fun reset()

    /**
     * Check if system is running
     */
    external fun isRunning(): Boolean

    /**
     * Convenience wrapper used by the foreground service so the call site
     * is symmetric with the rest of the API even though the JNI signature
     * matches `processSensor` underneath.
     */
    fun pushSensorEvent(sensorType: String, timestampMs: Long, values: FloatArray): Boolean {
        val typeId = when (sensorType) {
            "accelerometer" -> 1
            "gyroscope"     -> 4
            "magnetometer"  -> 2
            "light"         -> 5
            "proximity"     -> 8
            else            -> 0
        }
        return processSensor(typeId, values, timestampMs * 1_000_000L, 3)
    }
}
